### PR TITLE
converge ucs_status for action in the end of callback

### PR DESCRIPTION
**Pull request by changchengx**
Signed-off-by: Changcheng Liu <jerrliu@nvidia.com>

## What
It's possible that data is splitted into small chunks during write/read operation. If the internal chunk write/read operation is executed without that ucs_status_t is UCS_OK, then the end of callback should take the internal failure case into consideration.

### CODE REVIEW

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -610,8 +610,8 @@ class P2pDemoCommon : public UcxContext {
     public:
         SendCompleteCallback(size_t buffer_size,
                              MemoryPool<SendCompleteCallback>& pool) :
-            _op_counter(NULL), _counter(0), _iov(NULL), _io_msg(NULL),
-            _pool(pool) {
+            _fin_status(UCS_OK), _op_counter(NULL), _counter(0), _iov(NULL),
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:613]: rename _fin_status to _status

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -623,6 +623,7 @@ class P2pDemoCommon : public UcxContext {
         }
 
         virtual void operator()(ucs_status_t status) {
+            _fin_status |= status;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:626]: if (_status == UCS_OK) {
  _status = status;
}

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -632,14 +633,15 @@ class P2pDemoCommon : public UcxContext {
             }
 
             if (_io_msg != NULL) {
-                (*_io_msg)(status);
+                (*_io_msg)(ucs_status_t(_fin_status));
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:636]: no need to add ucs_status_t()

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -780,11 +782,12 @@ class DemoServer : public P2pDemoCommon {
         }
 
         virtual void operator()(ucs_status_t status) {
+            _fin_status |= status;
             if (--_chunk_cnt > 0) {
                 return;
             }
 
-            if (status == UCS_OK) {
+            if (ucs_status_t(_fin_status) == UCS_OK) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:790]: no need ucs_status_t()

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -807,6 +810,7 @@ class DemoServer : public P2pDemoCommon {
         }
 
     private:
+        uint32_t                             _fin_status;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:813]: ups_status_t _status

File: test/apps/iodemo/io_demo.cc (hunk 1 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 2 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: Done

File: test/apps/iodemo/io_demo.cc (hunk 3 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: cleaned

File: test/apps/iodemo/io_demo.cc (hunk 4 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: cleaned

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (changchengx) [reply test/apps/iodemo/io_demo.cc]: done

File: test/apps/iodemo/io_demo.cc [hunk 6]
```diff
@@ -1131,15 +1135,17 @@ class DemoClient : public P2pDemoCommon {
...
+            }
             if (--_comp_counter > 0) {
                 return;
             }
 
             assert(_server_index != std::numeric_limits<size_t>::max());
             _client->handle_operation_completion(_server_index, IO_READ);
 
-            if (_validate && (ucs_status_t(_fin_status) == UCS_OK)) {
+            if (_validate && _status == UCS_OK) {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1148]: pls add ( ) :
`if (_validate && (_status == UCS_OK)) {`

**Comment by changchengx** [comment]
The CI errors are not related with this PR
```
2021-06-21T13:00:56.2691730Z [----------] 5 tests from rcx/test_ucp_peer_failure
2021-06-21T13:00:56.2694841Z [ RUN      ] rcx/test_ucp_peer_failure.zcopy/3 <rc_x/rma_fail_imm>
2021-06-21T13:00:57.9210524Z [     INFO ] < request 0x3bbadd0 completed with error Endpoint timeout >
2021-06-21T13:00:57.9211708Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-21T13:00:57.9768231Z [       OK ] rcx/test_ucp_peer_failure.zcopy/3 (1708 ms)
2021-06-21T13:00:57.9769026Z [ RUN      ] rcx/test_ucp_peer_failure.force_close/3 <rc_x/rma_fail_imm>
2021-06-21T13:00:59.8007749Z [     INFO ] < request 0x3bbb050 completed with error Request canceled >
2021-06-21T13:00:59.8008945Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-21T13:00:59.8917574Z [       OK ] rcx/test_ucp_peer_failure.force_close/3 (1915 ms)
2021-06-21T13:00:59.8919481Z [ RUN      ] rcx/test_ucp_peer_failure.rndv_disable/3 <rc_x/rma_fail_imm>
2021-06-21T13:01:00.0990450Z [       OK ] rcx/test_ucp_peer_failure.rndv_disable/3 (207 ms)
2021-06-21T13:01:00.0991228Z [ RUN      ] rcx/test_ucp_peer_failure.basic/3 <rc_x/rma_fail_imm>
2021-06-21T13:01:01.9469570Z [     INFO ] < request 0x3bbb050 completed with error Endpoint timeout >
2021-06-21T13:01:01.9470314Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-21T13:01:02.0110945Z [       OK ] rcx/test_ucp_peer_failure.basic/3 (1912 ms)
2021-06-21T13:01:02.0112331Z [ RUN      ] rcx/test_ucp_peer_failure.bcopy_multi/3 <rc_x/rma_fail_imm>
2021-06-21T13:01:03.8269204Z [     INFO ] < request 0x3bbb050 completed with error Endpoint timeout >
2021-06-21T13:01:03.8270265Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-21T13:01:03.8840203Z [       OK ] rcx/test_ucp_peer_failure.bcopy_multi/3 (1873 ms)
2021-06-21T13:01:03.8841896Z [----------] 5 tests from rcx/test_ucp_peer_failure (7615 ms total)
```

**Comment by changchengx** [comment]
The CI errors are not related with this PR
```
2021-06-22T01:45:33.1192376Z [----------] 27 tests from dcudx/test_ucp_sockaddr_protocols_err
2021-06-22T01:45:33.1193707Z [ RUN      ] dcudx/test_ucp_sockaddr_protocols_err.tag_eager_32_unexp/1 <dc_x,ud,cuda_copy,rocm_copy/send_stop,mt>
2021-06-22T01:45:33.2813944Z [     INFO ] server listening on 12.10.44.11:53016
2021-06-22T01:45:33.4677478Z [     INFO ] ignoring error Connection reset by remote peer on endpoint 0x7f5d60146130
2021-06-22T01:45:33.4835283Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-22T01:45:33.5413759Z [       OK ] dcudx/test_ucp_sockaddr_protocols_err.tag_eager_32_unexp/1 (423 ms)
2021-06-22T01:45:33.5415604Z [ RUN      ] dcudx/test_ucp_sockaddr_protocols_err.tag_zcopy_64k_unexp/9 <dc_x,ud,cuda_copy,rocm_copy/bidi_stop,mt>
2021-06-22T01:45:33.7013683Z [     INFO ] server listening on 12.10.44.11:49699
2021-06-22T01:45:33.8878818Z [     INFO ] ignoring error Connection reset by remote peer on endpoint 0x7f5d60324130
2021-06-22T01:45:33.9580313Z [       OK ] dcudx/test_ucp_sockaddr_protocols_err.tag_zcopy_64k_unexp/9 (416 ms)
2021-06-22T01:45:33.9582176Z [ RUN      ] dcudx/test_ucp_sockaddr_protocols_err.tag_rndv_unexp/1 <dc_x,ud,cuda_copy,rocm_copy/send_stop,mt>
2021-06-22T01:45:34.1197518Z [     INFO ] server listening on 12.10.44.11:56358
2021-06-22T01:45:34.3058649Z [     INFO ] ignoring error Connection reset by remote peer on endpoint 0x7f5d60146130
2021-06-22T01:45:34.3216695Z [     INFO ] < request 0x319ed90 completed with error Request canceled >
2021-06-22T01:45:34.3218173Z [     INFO ] < request 0x43ae0d0 completed with error Request canceled >
2021-06-22T01:45:34.3222567Z [     INFO ] < request 0x319ed90 completed with error Endpoint timeout >
2021-06-22T01:45:34.3863769Z [       OK ] dcudx/test_ucp_sockaddr_protocols_err.tag_rndv_unexp/1 (428 ms)
2021-06-22T01:45:34.3865194Z [ RUN      ] dcudx/test_ucp_sockaddr_protocols_err.tag_zcopy_4k_unexp/5 <dc_x,ud,cuda_copy,rocm_copy/recv_stop,mt>
2021-06-22T01:45:34.5516943Z [     INFO ] server listening on 12.10.44.11:41828
2021-06-22T01:45:34.7367235Z [     INFO ] ignoring error Connection reset by remote peer on endpoint 0x7f5d60324130
2021-06-22T01:45:34.7523412Z [     INFO ] < operation returned error: Endpoint timeout >
```

**Comment by yosefe** [comment]
The failures in io_demo seem related

**Comment by changchengx** [comment]
@yosefe I'll check the io_demo failure case and feedback to you later.

**Comment by changchengx** [comment]
The failure of ```io_demo: tag``` [CI test case](https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/19991/logs/242) is because:
[iodemo_analyzer.py](https://github.com/openucx/ucx/blob/master/buildlib/io_demo/iodemo_analyzer.py) found that it has passed more than [1 minutes](https://github.com/openucx/ucx/blob/master/buildlib/io_demo/iodemo_analyzer.py#L220) from the time that the client got last completed read operation to the time that client got none completed read operation.
Client tried to set up 5 connection after can't get the requested window size io_msg again and again within the ```client_timeout``` time. Server doesn't reply IO_READ_COMP to client in time.

```
2021-06-22T01:21:24.1442777Z sputnik2: [1624324884.143244] [DEMO] read 2430.82 MB/s min:19493 (2.1.4.1:20000) max:19493 total:19493 ops, write 2426.43 MB/s min:19487 (2.1.4.1:20000) max:19487 total:19487 ops, connections:1/1, buffers:16
......
2021-06-22T01:22:52.1799074Z sputnik2: [1624324972.178964] [DEMO] read 0.0243761 MB/s min:0 (2.1.4.1:20000) max:0 total:3 ops, write 0.029769 MB/s min:0 (2.1.4.1:20000) max:0 total:7 ops, connections:0/0, buffers:16
```

**Comment by changchengx** [comment]
@yosefe I've found that
The internal _status could be ```UCS_ERR_CANCELED``` or ```UCS_ERR_ENDPOINT_TIMEOUT``` or ```UCS_ERR_CONNECTION_RESET```, but the final call to ```operator()(ucs_status_t status)``` has the argument value UCS_OK.

**Comment by changchengx** [comment]
@yosefe The ```io_demo: tag``` & ```io_demo: active``` test cases have passed.
The two failed ```Tests gpu on worker 0``` & ```Tests gpu on worker 2``` test cases are not related with this PR:
**Failed Tests gpu on worker 0**
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/20164/logs/356
```
021-06-24T08:26:48.8757096Z [----------] 5 tests from rc/test_ucp_peer_failure
2021-06-24T08:26:48.8757706Z [ RUN      ] rc/test_ucp_peer_failure.zcopy/0 <rc_v/am>
2021-06-24T08:26:50.6013012Z [     INFO ] < request 0x3e77350 completed with error Endpoint timeout >
2021-06-24T08:26:50.6014063Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-24T08:26:50.6465936Z [       OK ] rc/test_ucp_peer_failure.zcopy/0 (1772 ms)
2021-06-24T08:26:50.6467341Z [ RUN      ] rc/test_ucp_peer_failure.force_close/0 <rc_v/am>
2021-06-24T08:26:52.2133524Z [     INFO ] < request 0x3a3b6d0 completed with error Request canceled >
2021-06-24T08:26:52.2707025Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-24T08:26:52.2708045Z [       OK ] rc/test_ucp_peer_failure.force_close/0 (1624 ms)
2021-06-24T08:26:52.2709656Z [ RUN      ] rc/test_ucp_peer_failure.rndv_disable/0 <rc_v/am>
2021-06-24T08:26:52.4348285Z [       OK ] rc/test_ucp_peer_failure.rndv_disable/0 (164 ms)
2021-06-24T08:26:52.4349815Z [ RUN      ] rc/test_ucp_peer_failure.basic/0 <rc_v/am>
2021-06-24T08:26:54.0935256Z [     INFO ] < request 0x3e775d0 completed with error Request canceled >
2021-06-24T08:26:54.0936330Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-24T08:26:54.1302516Z [       OK ] rc/test_ucp_peer_failure.basic/0 (1695 ms)
2021-06-24T08:26:54.1303989Z [ RUN      ] rc/test_ucp_peer_failure.bcopy_multi/0 <rc_v/am>
2021-06-24T08:26:55.7028689Z [     INFO ] < request 0x3e775d0 completed with error Request canceled >
2021-06-24T08:26:55.7401850Z [     INFO ] < operation returned error: Endpoint timeout >
2021-06-24T08:26:55.7402817Z [       OK ] rc/test_ucp_peer_failure.bcopy_multi/0 (1610 ms)
2021-06-24T08:26:55.7405263Z [----------] 5 tests from rc/test_ucp_peer_failure (6865 ms total)
```

**Failed Tests gpu on worker 2**
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=20164&view=logs&jobId=73bd2a9e-6688-5c22-d122-7b1a0b6cbb81
```
2021-06-24T08:12:42.2982665Z + taskset -c 14 ./examples/uct_hello_world -d mlx5_1:1 -t rc_verbs -z -m cuda-managed -p 18016
2021-06-24T08:12:42.3176474Z INFO: UCT_HELLO_WORLD AM function = uct_ep_am_zcopy server = (null) port = 18016
2021-06-24T08:12:42.3471623Z Unsupported memory type cuda-managed by rc_verbs/mlx5_1:1 on mlx5_1 MD
2021-06-24T08:12:42.3483428Z No supported (dev/tl) found (mlx5_1:1/rc_verbs)
2021-06-24T08:12:42.3491911Z Failed to find supported device and transport
2021-06-24T08:12:57.2972583Z + '[' 0 -eq 1 ']'
2021-06-24T08:12:57.2973926Z + client_pid=20735
2021-06-24T08:12:57.2974649Z + wait 20735
2021-06-24T08:12:57.2986794Z + taskset -c 15 ./examples/uct_hello_world -d mlx5_1:1 -t rc_verbs -z -m cuda-managed -n swx-rdmz-ucx-gpu-01 -p 18016
2021-06-24T08:12:57.3189959Z INFO: UCT_HELLO_WORLD AM function = uct_ep_am_zcopy server = swx-rdmz-ucx-gpu-01 port = 18016
2021-06-24T08:12:57.3528006Z Unsupported memory type cuda-managed by rc_verbs/mlx5_1:1 on mlx5_1 MD
2021-06-24T08:12:57.3536247Z No supported (dev/tl) found (mlx5_1:1/rc_verbs)
2021-06-24T08:12:57.3537315Z Failed to find supported device and transport
```

**Review by yosefe — APPROVED** [review]
pls squash (this will also trigger retest)

**Review by yosefe — APPROVED** [review]

**Pull request merged**